### ISSUE TITLE
replaceState for blocks not working properly anymore

### ISSUE THREAD

**Issue by MarvinVrdoljak** [payloadcms/payload#2854]
### Link to reproduction

-

### To Reproduce

- Insert code above in your project
- log in into you payload admin area and go to my test-collection "Tests"
- add new "Test"
- click on the button "click" in sidebar
-> New State of Title is visible; new state of array is NOT

Note: This Code is working fine in version 1.8.5

**Collection:**
```js
import {CollectionConfig} from 'payload/types'
import {ReplaceStateComponent} from '../component/collectionComponents/ReplaceStateComponent'

export const CollectionTest: CollectionConfig = {
  slug: 'test',
  admin: {
    useAsTitle: 'title',
    defaultColumns: ['fullTitle', 'author', 'createdAt', 'status'],
  },
  versions: {
    drafts: true,
  },
  fields: [
    {
      name: 'title',
      type: 'text',
      required: true,
    },
    {
      name: 'array',
      type: 'blocks',
      blocks: [
        {
          slug: 'text',
          fields: [
            {
              name: 'text',
              type: 'text',
              required: true,
            },
          ],
        },
        {
          slug: 'textarea',
          fields: [
            {
              name: 'textarea',
              type: 'textarea',
              required: true,
            },
          ],
        },
      ],
    },
    {
      name: 'preview',
      type: 'ui',
      admin: {
        position: 'sidebar',
        components: {
          Field: ReplaceStateComponent,
        },
      },
    },
  ],
}
```

Component: 
```js
import React from 'react'
import {useForm} from 'payload/components/forms'

interface StateValue {
  initialValue: any
  value: any
  valid: boolean
  disableFormData?: boolean
}

type State = Record<string, StateValue>

export const ReplaceStateComponent: React.FC = () => {
  const {getFields, replaceState, setModified} = useForm()
  const fields = getFields()

  const importBlocks = async (e: React.MouseEvent<HTMLElement>) => {
    e.preventDefault()

    const newState: State = {...fields}

    newState['title'] = {
      initialValue: 'title',
      value: 'title',
      valid: true,
      disableFormData: true,
    }

    newState['array'] = {
      initialValue: 1,
      value: 1,
      valid: true,
      disableFormData: true,
    }

    newState['array.0.blockName'] = {
      initialValue: 'textarea',
      value: 'textarea',
      valid: true,
    }

    newState['array.0.blockType'] = {
      initialValue: 'textarea',
      value: 'textarea',
      valid: true,
    }

    newState['array.0.id'] = {
      initialValue: '64904e5c8714d9702677551c',
      value: '64904e5c8714d9702677551c',
      valid: true,
    }

    newState['array.0.textarea'] = {
      initialValue: 'This is a test',
      value: 'This is a test',
      valid: true,
    }

    // Sets the new state
    replaceState(newState)
    setModified(true)

   // setTimeout(() => {
   //  submit()
   //  }, 1000)
  }

  return (
    <button type="button" onClick={importBlocks}>
      Click
    </button>
  )
}
```

### Describe the Bug

The code (example) was running perfectly fine in version 1.8.5.
After clicking the "click"-button you need to submit the collection-post manually or use the submit() function at the end of the code to see the new state of the array.

### Payload Version

1.9.5

**Comment by jmikrut** [comment]
Hey @MarvinVrdoljak — I can shine some light on this for you.

Updating array rows is a not yet documented feature, but we are working on making it significantly easier.

TL;DR - we used to store _**row state**_ separately in its own state, per row field, which is what made updating rows of array / block fields difficult.

We now have flattened row state to exist in the parent form state itself, which will allow us to expose simple helper functions like `addRow` and `removeRow` directly from form state. That will simplify the work involved here significantly.

Right now, you can leverage `dispatchFields` by sending an `ADD_ROW` action to add rows to an array field, but, easier methods will be coming shortly like I mentioned. 

You can also simply adjust your code by adding rows into your existing `replaceState` function by updating your `array` field to the following:

```js
    newState.array = {
      initialValue: 1,
      value: 1,
      valid: true,
      disableFormData: true,

      // NEWLY CONSOLIDATED ROWS PROPERTY
      rows: [
        {
          blockType: 'textarea',
          collapsed: false,
          id: '64904e5c8714d9702677551c',
        },
      ],
    };
```

Give that a shot, will you?

Sorry about the hassle - this will all become officially supported, and documented, shortly! We're making a lot of progress here.

**Comment by MarvinVrdoljak** [comment]
@jmikrut Thanks a lot for your reply. Do I understand correctly that I can only use the row property to add empty rows? Because I actually need to add them prefilled. This is how my code looks like now: 

```js
    newState['array'] = {
      initialValue: 1,
      value: 1,
      valid: true,
      disableFormData: true,
      rows: [
        {
          blockType: 'textarea',
          collapsed: false,
          id: '64904e5c8714d9702677551c',
          // textarea: 'textarea', // Not working -> Would be great if this would work
          // blockName: 'textarea', // Not working -> Would be great if this would work
        },
      ],
    }

    // To fill the added rows with content:
    newState['array.0.textarea'] = {
      initialValue: 'textarea',
      value: 'textarea',
      valid: true,
    }

    newState['array.0.blockName'] = {
      initialValue: 'textarea',
      value: 'textarea',
      valid: true,
    }
```

I really appreciate and looking forward to your official solutions regarding this topic. Can you already estimate when you will release it?

**Comment by JarrodMFlesch** [comment]
@MarvinVrdoljak what you have above is correct, to replace state you will need to add the rows individually like you are doing with `newState['array.0.textarea']`.

With the new approach you would need to implement something similar to this:

https://github.com/payloadcms/payload/blob/master/src/admin/components/forms/field-types/Array/index.tsx#L108

The helper functions will likely just abstract this out and have you pass `rowIndex`, `path` and `fields` so we can do the `buildStateFromSchema` part for you. As for ETA, not sure yet. I will talk with the team!

**Comment by taismassaro** [comment]
hey @JarrodMFlesch, do you have any updates on this? I've got an Article collection that has an array field for article sections and I would like to prefill some section titles when creating a new article. Ideally this would be done based on a previously chosen article type, but for now I'm just trying to prefill one "Summary" section in any new article. I got it somewhat working following the code in this issue, but it's very unclear to me what needs to go in the `rows` array and I'm getting a console error:
```
Uncaught TypeError: Cannot read properties of undefined (reading 'size')
    at setsAreEqual (setsAreEqual.js:4:1)
    at index.js:120:1
    at Array.every (<anonymous>)
    at index.js:117:1
    at Array.forEach (<anonymous>)
    at index.js:116:1
    at WatchFormErrors.js:10:1
    at useThrottledEffect.js:10:1
```

this is the code I have for the UI field:

```ts
const { getFields, replaceState, setModified } = useForm();

  const fields = getFields();
  console.log({ fields });

  useEffect(() => {
    if (fields.sections.rows.length <= 0) {
      const updatedFields = {
        ...fields,
        sections: {
          ...fields.sections,
          disableFormData: true,
          rows: [
            {
              id: "64904e5c8714d9702677551c",
              content: "richText",
            },
          ],
        },
        "sections.0.title": {
          initialValue: "Summary",
          value: "Summary",
          valid: true,
        },
      };

      replaceState(updatedFields);
      setModified(true);
    }
  }, []);
```

any insights would be appreciated! 🙏🏽
